Bug 1634059 - copyedit of "Getting Started" sections (#472)
mozilla/firefox-data-docs: src/tools/stmo.md

Introduction to STMO
--------------------

[`sql.telemetry.mozilla.org`](https://sql.telemetry.mozilla.org) (STMO) is an outstanding [Re:dash](https://redash.io/) data analysis and dashboarding tool. It has been customized and configured for use with a number of the Firefox organization's data sets. As the name and URL imply, the effective use of this tool requires familiarity with
[SQL](https://en.wikipedia.org/wiki/SQL), with which all of the
tool's data extraction and analysis are performed.

Access to STMO is limited to Mozilla employees and designated contributors. For more informationn, see the [gaining access](../concepts/gaining_access.md) document.

> STMO Concepts

Author: Can we pls implement this in the second phase?

---------------------------

The following topics describe the process of creating a simple dashboard using STMO.

#### Create A Query

You start by creating a query. The first query counts the number of client ids that Mozilla receives from each country, for the top N countries. If you click the 'New Query' button that is located at the top on the left-hand side of the site, the query editing page appears:

![New Query Page](../assets/STMO_screenshots/new_query.png "New Query page")

For this and most other queries where each client IDs is counted, you want to use
[`clients_last_seen`](../datasets/bigquery/clients_last_seen/reference.md),
which Firefox telemetry pings generate.

* Search for the table in `Telemetry (BigQuery)`

  Click 'Data Source' drop-down and select `Telemetry (BigQuery)`. Then search for the table by typing `clients_last_seen` in the "Search schema..." Search field that is above the schema browser interface to the left of the main query Edit field.
  
  A `clients_last_seen` table appears as `telemetry.clients_last_seen`). Versioned `clients_last_seen` data sets also appear as `telemetry.clients_last_seen_v<VERSION>`).

* Unable to locate the table

> 

Author: Can we pls delete this in a second phase?

  mau DESC
```

> If you type these parameters into the main query Edit field and then click the "Execute"
> button, a blue bar then appears below the Edit field. It includes the text "Executing query...", followed by a timer that indicates how long the query has run. After a reasonable period of time, usually less than a minute, the query typically completes its run. A table appears that displays a MAU value for each country. You have just created and run your first STMO query.

Reviewer: I think this is one of those rare cases we can use an exclamation mark, since the user has accomplished something significant by this point in the tutorial.

```suggestion
If you type these parameters into the main query Edit field and then click the "Execute"
button, a blue bar then appears below the Edit field. It includes the text "Executing query...", followed by a timer that indicates how long the query has run. After some period of time, usually less than a minute, the query typically completes its run. A table appears that displays a MAU value for each country. You have just created and run your first STMO query!
```

Next, click the large "New Query" text located at the top of the page. An Edit field appears so you can rename the query. It is recommended that you assign a unique prefix (such as your name) to the query to make it easy to find your query later. For example, 
`rmiller:Top Countries`.

#### Create A Visualization

> After you have created a query, you need to provide a simple visualization. The table with results from the first query execution now appears under the query Edit field. Another heading titled `+NEW VISUALIZATION` appears bext to the `TABLE` heading:

Reviewer: ```suggestion
After you have created a query, you may want to provide a simple visualization. The table with results from the first query execution now appears under the query Edit field. Another heading titled `+NEW VISUALIZATION` appears next to the `TABLE` heading:
```

  20
```

> If you edit the query to add a limit clause and then click 'Execute', a new bar graph only displays the 20 countries that have the highest number of unique clients. In this case, the full result set includes approximately 250 return values. Limiting the result count improves readability.

Reviewer: ```suggestion
If you edit the query to add a limit clause and then click 'Execute', a new bar graph only displays the 20 countries that have the highest number of unique clients. In this case, the full result set includes approximately 250 return values: limiting the result count improves readability.
```

> In other cases, however, an unlimited query may return thousands or even millions of rows. When these queries are run, readability is not the only problem that occurs. Any queries that return millions of rows or lines can fail to achieve the results and negatively impact the performance of all other users of a system. Make sure that you include:

Reviewer: ```suggestion
In other cases, however, an unlimited query may return thousands or even millions of rows. Any queries that return millions of rows or lines can not only be unreadable but negatively impact the performance of all other users of STMO. Thus an important warning:
```

  **ALL QUERIES SHOULD INCLUDE A "LIMIT" STATEMENT BY DEFAULT!**

It is highly recommended that you add a "LIMIT 100" clause to the end of all new queries to prevent a query from returning a large result set that causes user interface (UI) and performance problems. You may learn that the total result set is small enough that setting a limit becomes unnecessary. Specifying an explicit LIMIT helps prevent unnecessary issues.

#### Query Parameters

> Adding a "LIMIT 20" clause at the end of a query makes a chart more manageable. If you only want to view the first 10 or first 30 entries, you can modify queries to accept user arguments by setting double curly-braces around a variable name, as shown below:
> 

Reviewer: ```suggestion
You can add user arguments to a query, which allows the user to specify parameters without modifying the query itself. Using the query in the previous section as an example, we can replace the `LIMIT 20` with a country count variable in double curly braces:
```

Author: I would replace "we" with "you"

  {{country_count}}
```

> After you have replaced the hard-coded limit value with `{{country_count}}`, an input field appears above the bar chart. If you type a numeric value in the input field and click 'Execute', the query runs with the specified limit. Click 'Save' to save the query. The query applies the parameter value that you typed as the default.

Reviewer: ```suggestion
After you have replaced the hard-coded limit value with `{{country_count}}`, an input field appears above the bar chart. If you type a numeric value in the input field and click 'Execute', the query will run with the specified limit. Click 'Save' to save the query. The query applies the parameter value that you typed as the default.
```

#### Create A Dashboard

> You can create a dashboard to display visualization by selecting 'New Dashboard' from the 'Dashboards' dropdown that is located at the top left of the page. Choose a name for your dashboard and an almost empty page appears. Click the '...' button near the top right of the page to selectn'Add Widget'. The 

Reviewer: ```suggestion
Now we can create a dashboard to display our visualization. Do this by selecting 'New Dashboard' from the 'Dashboards' dropdown that is located at the top left of the page. Choose a name for your dashboard and an empty page will appear. Click the '...' button near the top right of the page, and you will see the option to 'Add Widget'. Click that, and the 

```

Author: I would remove the future tense "will". Also, who is "we"? Mozilla?

Reviewer: Agreed, reverting back to (mostly) your formulation, with a few small typos fixed

> following dialog box appears:

Reviewer: ```suggestion
following dialog box will appear:
```

![Add Widget](../assets/STMO_screenshots/add_widget.png "Add Widget")

> Type a unique prefix that has previously been used in a query name in the "Search a query by name" field to locate the query that you have created. You cannot yet execute this query because the query has not yet been published. As soon as you publish a query, you can search on the summary pages. Even though this is only an exercise, the query must be published briefly and then added to the dashboard. You can publish your query by clicking "Publish" on the query source page.

Reviewer: I think some of the meaning and explanation was lost here-- I think it might be best to revert this back to the original text, as I think it's relatively easy to understand.

Author: Not sure how best to do this - can we please discuss

As soon as a query is published, it appears in the search results when you type a unique prefix in the "Search a query by name" field on the "Add Widget" dialog box. When you select a query, you can select a query's visualizations from the "Choose Visualization" dropdown. Select the bar chart that you created and then click "Add to Dashboard". The dashboard now includes a bar chart. You can now edit the `country_count` value and click "Refresh" to change the number of countries that are included in the chart.

#### Completing the Dashboard

> A dashboard with just one graph is often insufficient. Therefore, you may want to create additional  queries, each with a very similar bar chart. The text that you need to apply to the queries is listed below. However, you need to create the queries and the visualizations and then link them to the dashboard. The queries are as follows:

Reviewer: ```suggestion
A dashboard with just one graph is often insufficient. Therefore, you may want to create additional queries, each with a very similar bar chart. The text that you need to apply to the queries is listed below. However, you need to create the queries and the visualizations and then link them to the dashboard. The queries are as follows:
```

can result in a dashboard that resembles this:

![Completed Dashboard](../assets/STMO_screenshots/finished_dashboard.png "Completed Dashboard")

Some final notes to help you create your dashboards:

* Remember to publish each of your queries before adding its visualizations to a dashboard.

* Similarly, it is recommended to un-publish any test queries after you have used
  them in a dashboard. This prevents everyone's search results from being contaminated
  with your tests and experiments. Any queries that result from an actual
  work-related analysis typically remain published. Others users can view these queries and
  learn from them.

* The 'Firefox' label on the 'App Version counts' graph is related to the use
  of the 'Group by' visualization setting. It is recommended that you experiment with
  the use of 'Group by' in your graphs to learn more about its usage.

* This tutorial has only touched the surface of the wide variety of
  sophisticated visualizations that STMO supports. You can view many
  more sophisticated queries and dashboards by browsing and exploring
  the work that has been published by others.

> * The [Re:dash help center](https://redash.io/help/) is helpful if you want to explore 
>   Re:dash and all of its capabilities further.

Reviewer: ```suggestion
* The [Re:dash help center](https://redash.io/help/) is a useful resource for exploring Re:dash and all its capabilities.
```

> #### Prototype Queries

Reviewer: I think the original wording is a bit more clear

```suggestion
#### Prototyping Queries
```

You may want to start working on a query before data becomes available.
You can do this with most of the data sources by selecting a static test data
set and then work with it, as usual. You can also use this method to explore
how a particluar SQL backend behaves.

> Note that `UNION ALL` retain duplicate rows while `UNION` deletes them.

Reviewer: ```suggestion
Note that `UNION ALL` will retain duplicate rows while `UNION` will discard them.
```
